core/src/main/java/org/apache/iceberg/avro/ValueReaders.java

```diff
@@ -566,8 +569,10 @@ public Map<K, V> read(Decoder decoder, Object reuse) throws IOException {
 
   public abstract static class StructReader<S> implements ValueReader<S>, SupportsRowPosition {
     private final ValueReader<?>[] readers;
-    private final int[] positions;
-    private final Object[] constants;
+    private final int[] constantPositions;
+    private final Object[] constantValues;
+    private final int[] defaultPositions;
+    private final Object[] defaultValues;
```

Comment:
If we go with the approach in #9366, then we should no longer need changes here, which was what originally made this difficult.